qt/aqt/import_export/exporting.py

```diff
@@ -0,0 +1,219 @@
+
+    def current_deck(self) -> DeckNameId | None:
+        if self.exporter.show_deck_list:
+            if idx := self.frm.deck.currentIndex():
+                return self.all_decks[idx - 1]
+        return None
+
+    def filename(self) -> str:
+        if self.exporter.show_deck_list:
+            deck_name = self.frm.deck.currentText()
+            stem = re.sub('[\\\\/?<>:*|"^]', "_", deck_name)
+        else:
+            time_str = time.strftime("%Y-%m-%d@%H-%M-%S", time.localtime(time.time()))
+            stem = f"{tr.exporting_collection()}-{time_str}"
+        return f"{stem}.{self.exporter.extension}"
+
+
+@dataclass
+class Options:
+    out_path: str
+    include_scheduling: bool
+    include_media: bool
+    include_tags: bool
+    include_html: bool
+    deck_id: DeckId | None
+    note_ids: Sequence[NoteId] | None
+
+
+class Exporter(ABC):
+    extension: str
+    show_deck_list = False
+    show_include_scheduling = False
+    show_include_media = False
+    show_include_tags = False
+    show_include_html = False
+
+    @staticmethod
+    @abstractmethod
+    def export(mw: aqt.main.AnkiQt, options: Options) -> None:
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def name() -> str:
+        pass
+
+
+class ColpkgExporter(Exporter):
+    extension = "colpkg"
+    show_include_media = True
+
+    @staticmethod
+    def name() -> str:
+        return tr.exporting_anki_collection_package()
+
+    @staticmethod
+    def export(mw: aqt.main.AnkiQt, options: Options) -> None:
+        def on_success(_future: Future) -> None:
+            mw.reopen()
+            tooltip(tr.exporting_collection_exported())
+
+        def on_failure(exception: Exception) -> None:
+            mw.reopen()
+            show_exception(parent=mw, exception=exception)
+
+        gui_hooks.collection_will_temporarily_close(mw.col)
+        QueryOpWithBackendProgress(
+            parent=mw,
+            op=lambda col: col.export_collection_package(
+                options.out_path, include_media=options.include_media, legacy=False
+            ),
+            success=on_success,
+            key="exporting",
+        ).failure(on_failure).run_in_background()
+
+
+class ApkgExporter(Exporter):
+    extension = "apkg"
+    show_deck_list = True
+    show_include_scheduling = True
+    show_include_media = True
+
+    @staticmethod
+    def name() -> str:
+        return tr.exporting_anki_deck_package()
+
+    @staticmethod
+    def export(mw: aqt.main.AnkiQt, options: Options) -> None:
+        QueryOpWithBackendProgress(
+            parent=mw,
+            op=lambda col: col.export_anki_package(
+                out_path=options.out_path,
+                selector=options.note_ids or options.deck_id or None,
+                with_scheduling=options.include_scheduling,
+                with_media=options.include_media,
+            ),
+            success=lambda fut: tooltip(tr.exporting_note_exported(count=fut)),
```

Comment:
The way QueryOpWithBackendProgress() is currently implemented unfortunately means we're losing typechecking here. 

- mypy has no way of knowing that the result of export_anki_package() will be passed to the success fn. You've labelled the argument 'fut' there, but it's actually an int. I think some code below is also incorrectly typed as a future, but we're not being notified about it currently.
- code completion/hovering in export_anki_package() is broken because the editor doesn't know that col is a Collection

I presume the current design was motivated by the fact that collection and query ops take different op callables, and you wanted to avoid code duplication. I think typechecking/code completion is more important here, and would be fine with some of that code being duplicated if it meant typechecking worked like standard collection/query ops.

Rather than subclassing the original classes, would it make sense to enable this additional behaviour via an extra argument to the original CollectionOp/QueryOp classes instead? Eg if a progress_key arg were provided, the timer could be started up, and ._run() could take a different codepath. Perhaps some/all of OpWithBackendProgress._run() could be turned into a free-function that both QueryOp and CollectionOp could call?

One other point to note is that currently we need to teach OpWithBackendProgress() about each new progress type we want to handle. The approach I took in AnkiMobile instead delegates extraction/label updating to the calling code. Eg, instead of passing in a label directly, it passes in a function that extracts the data it wants, and if it finds it, it calls the equivalent of mw.progress.update(...) itself. That way the calling code has control over how the data is rendered, and the code related to a specific operation lies in the file dealing with that operation, instead of in our shared operations code.

Reply:
You're right, and it's a bad use of inheritance. One more drawback: The signature of the child classes is rather unhelpful, because they have to accept arbitrary args and kwargs.
Implementing this in a free function is definitely possible (if state is maintained in a returned closure or so), but there already is a class `ProgressHandler` that we call. Would make it sense to place the backend progress logic there as well?

Reply:
Do you mean ProgressManager? I only see ProgressHandler in our Rust code. If you meant putting the generic backend progress handling code rather than area-specific code into ProgressManager, that's basically what I did in AnkiMobile, so it works for me :-) The AnkiMobile equiv looks roughly like this - a routine that the calling code can call after it calls progress.start():

```
    func monitorProgress(
        secs: Double,
        callback: (BackendProgress.OneOf_Value) -> Void
    ) {
        self.progressTimer = Timer.scheduledTimer(
            withTimeInterval: secs,
            repeats: true,
            block: { _ in
                // if there's an active progress value, send it to the callback, which will check it's the
                // type it expects, and then act on it if it is
                if let progressEnum = try? am.col.backend.latestProgress(),
                    let progress = progressEnum.value {
                    callback(progress)
                }
            }
        )
    }
```

The timer is automatically cleaned up on progress.finish().

Reply:
Yes, sorry for the typo. And thanks for the snippet!

Reply:
Good call, that was a big improvement. Do you still see any issues?

Reply:
Much nicer! 

Having the callback return a string is the most ergonomic in cases like this, but there are cases where we want to update the progress bar as well, such as dbcheck.py, and full syncs in sync.py. I'm not suggesting we rewrite those existing cases, just that it's a use case we might want to more explicitly support for future code. We could do this by having the callback be passed in the mw.progress instance, and have it make the update call itself, eg:

```python
# before
def export_progress_label(progress: Progress) -> str | None:
    if not progress.HasField("exporting"):
        return None
    return tr.exporting_exported_media_file(count=progress.exporting)

# after
def on_backend_progress(manager: ProgressManager, progress: Progress):
    if not progress.HasField("exporting"):
        return
    manager.update(label=tr.exporting_exported_media_file(count=progress.exporting))
```

Code could then pass in value= and max= if it wanted to update the progress bar.

We could theoretically do both - return a label, and allow the routine to also call .update() itself. That just feels a little less efficient and a bit less clear. WDYT?

Reply:
As your function doesn't return anything, can we `set_wants_abort()` then regardless of whether the closure processed the progress message?
Instead of passing the whole ProgressManager and expecting the user to call the right method, what do you think of taking the arguments for `update()`? As its signature is a bit cumbersome, we could accept a dataclass instead.

```py
def progress_update(progress: Progress) -> ProgressUpdate | None:
    if not progress.HasField("exporting"):
        return None
    return ProgressUpdate(label=tr.exporting_exported_media_file(count=progress.exporting))
```

Reply:
Don't have a strong preference here; returning a dataclass sounds fine too.

> can we set_wants_abort() then regardless of whether the closure processed the progress message?

At one point in the future we may want to handle multiple long-running ops at once (at the moment if a media sync is running at the same time, what get cancelled will depend on which progress happens to run next), and at that time we'll probably want to be only calling that in a routine-specific code (so that it can eg. pass an enum back to set_wants_abort() specifying what is to be aborted). So an argument could be made for moving the check out of progress.py and into the individual progress handlers instead, at the cost of some more work in the short term.

Reply:
Not sure I got this. Doesn't the GUI prevent the user from aborting any spcific op anyway, because it's only a single dialog?
If that's not relevant for this PR, feel free to just resolve this conversation.

Reply:
If some cancellable operation + media sync is running at the same time, closing the progress window will cancel one of the two operations - whichever happens to post progress next. As that's not ideal, we'll probably want to fix it at one point in the future. One possible way to do so would be to change set_wants_abort() to take an enum or string specifying the action that we want to cancel. I'm not suggesting we do that now, but if we leave the the set_wants_abort() call in start_with_backend_updates(), that's going to make it harder to pass an enum in the future. If instead progress_update() were passed a wants_abort, it could then call set_wants_abort() itself, after it's determined that the progress message is the one it's interested in. That wouldn't change anything at the moment, but would make it easier to pass in a relevant enum in the future.

Does that make sense? Alternative ideas welcome :-)

Reply:
Thanks, got it. I was referring to the fact that the GUI currently only has a single cancel signal, so the user can't decide which op to abort anyway, but I see that's not the point here.

As we're getting data back from the callback anyway, couldn't we just add a `ProgressUpdate.abort_variant` that gets passed on to `set_wants_abort()` if appropriate? Alternatively, `start_with_backend_updates()` could also take an `abort()` callback.
My issue with passing a `wants_abort` (or primitives in general) into the callback is, that it's hard to communicate its meaning. This may not be a problem in single-purpose callbacks, but here the type would be `Callable[[Progress, bool], ProgressUpdate | None]`. Then you'll have to explain in comments what the caller is supposed to do with that bool. (I guess you could use a newtype, but well...)

Reply:
The way I do it in AnkiMobile is pass in a mutable object that also contains the cancellation flag from the GUI, eg something like:

```python
    def progress_update(progress: Progress, update: ProgressUpdate) -> None:
        if not progress.HasField("exporting"):
            return None
        update.label = tr.exporting_exported_media_file(count=progress.exporting)
        if update.user_wants_abort:
            update.abort = True       
```

The invoking progress code then invokes set_wants_abort() if abort is true. That saves us from having to create an enum at the moment, and when we want to add one in the future, only the update.abort line + the set_wants_abort() invocation needs to be changed . Would that work for you?

> couldn't we just add a ProgressUpdate.abort_variant that gets passed on to set_wants_abort() if appropriate?

Unconditionally assigning it each time progress_update() is called feels a bit unclean, though the performance difference would be negligible.

Passing something into start_with_backend_updates() could work, but it might be slightly nicer to have all the logic inside the callback. WDYT?

Reply:
Yes, makes sense. I've pushed a commit.